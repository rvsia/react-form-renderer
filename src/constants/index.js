export const components = {
  TEXT_FIELD: 'text-field',
  TEXTAREA_FIELD: 'textarea-field',
  FIELD_ARRAY: 'field-array',
  SELECT_COMPONENT: 'select-field',
  FIXED_LIST: 'fixed-list',
  CHECKBOX: 'checkbox',
  SUB_FORM: 'sub-form',
  RADIO: 'radio',
  TABS: 'tabs',
  TAB_ITEM: 'tab-item',
  DATE_PICKER: 'date-picker',
  TIME_PICKER: 'time-picker',
  TAG_CONTROL: 'tag-control',
  WIZARD: 'wizard',
};
  /** Validator functions placeholder */
export const validators = {
  REQUIRED: 'required-validator',
  /**
     * min length of the input value
     */
  MIN_LENGTH: 'min-length-validator',
  /**
     * max length of the input value
     */
  MAX_LENGTH: 'max-length-validator',
  /**
   * exact length of input value
   */
  EXACT_LENGTH: 'exact-length-validator',
  /**
     * minimum count of fileds in some dynamic list of fields
     */
  MIN_ITEMS_VALIDATOR: 'min-items-validator',
  /**
     * Minimum value of number input
     */
  MIN_NUMBER_VALUE: 'min-number-value',
  /**
     * Maximum value of number inpuy
     */
  MAX_NUMBER_VALUE: 'max-number-value',
  /**
     * Regexp pattern validator
     */
  PATTERN_VALIDATOR: 'pattern-validator',
};

export const layoutComponents = {
  FORM_WRAPPER: 'FormWrapper',
  BUTTON: 'Button',
  COL: 'Col',
  FORM_GROUP: 'FormGroup',
  BUTTON_GROUP: 'ButtonGroup',
  ICON: 'Icon',
  ARRAY_FIELD_WRAPPER: 'ArrayFieldWrapper',
  HELP_BLOCK: 'HelpBlock',
};
