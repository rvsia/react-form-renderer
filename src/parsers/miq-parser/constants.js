import { components } from '../../constants';

export const componentMap = {
  DialogFieldTextBox: components.TEXT_FIELD,
  DialogFieldRadioButton: components.RADIO,
  DialogFieldCheckBox: components.CHECKBOX,
  DialogFieldTextAreaBox: components.TEXTAREA_FIELD,
  DialogFieldDropDownList: components.SELECT_COMPONENT,
  DialogFieldDateControl: components.DATE_PICKER,
  DialogFieldDateTimeControl: components.TIME_PICKER,
  DialogFieldTagControl: components.TAG_CONTROL,
};

// for true values
export const neededAttributes = [
  'name',
  'label',
  [ 'label', 'title' ],
  [ 'data_type', 'dataType' ],
  [ 'required', 'isRequired' ],
  [ 'read_only', 'isReadOnly' ],
  [ 'description', 'helperText' ],
];

// for false values
export const neededFalseAttributes = [
  [ 'visible', 'isVisible' ],
];
