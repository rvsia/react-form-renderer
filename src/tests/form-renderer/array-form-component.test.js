import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import { Form } from 'react-final-form';
import arrayMutators from 'final-form-arrays';
import renderForm from '../../form-renderer/render-form';
import RendererContext, { configureContext } from '../../form-renderer/renderer-context';
import { components } from '../../constants';
import FormRenderer from '../../form-renderer';
import layoutMapper from '../../../demo/layout-mapper';

describe('renderForm function', () => {
  let formFieldsMapper;

  const ContextWrapper = ({ initialValues = {}, ...rest }) => (
    <FormRenderer
      formFieldsMapper={ formFieldsMapper }
      layoutMapper={ layoutMapper }
      initialValues={ initialValues }
      onSubmit={ jest.fn() }
      { ...rest }
    />
  );

  beforeEach(() => {
    formFieldsMapper = {
      [components.TEXT_FIELD]: () => <div className="nested-item">Text field</div>,
    };
  });

  it('should render array field correctly', () => {
    const formFields = [{
      component: components.FIELD_ARRAY,
      name: 'foo',
      key: 'bar',
      fields: [{
        component: components.TEXT_FIELD,
        name: 'nested component',
        label: 'foo',
      }],
    }];

    const wrapper = mount(
      <ContextWrapper schema={{ fields: formFields }}/>
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should add nested field to form', () => {
    const formFields = [{
      component: components.FIELD_ARRAY,
      name: 'foo',
      key: 'bar',
      fields: [{
        component: components.TEXT_FIELD,
        name: 'nested component',
        label: 'foo',
      }],
    }];

    const wrapper = mount(
      <ContextWrapper schema={{ fields: formFields }} />
    );

    expect(wrapper.find(Form).instance().form.getState().values.foo).toBeUndefined();
    wrapper.find('button').first().simulate('click');
    wrapper.update();
    expect(wrapper.find(Form).instance().form.getState().values.foo).toHaveLength(1);
  });

  it('should remove nested field to form', () => {
    const formFields = [{
      component: components.FIELD_ARRAY,
      name: 'foo',
      key: 'bar',
      fields: [{
        component: components.TEXT_FIELD,
        name: 'nested component',
        label: 'foo',
      }],
    }];

    const wrapper = mount(
      <ContextWrapper initialValues={{ foo: [ 'first value' ]}} schema={{ fields: formFields }}/>
    );
    expect(wrapper.find(Form).instance().form.getState().values.foo).toHaveLength(1);
    wrapper.find('button').first().simulate('click');
    wrapper.update();
    expect(wrapper.find(Form).instance().form.getState().values.foo).toBeUndefined();
  });

  it('should render fixed array field correctly', () => {
    const formFields = [{
      component: components.FIXED_LIST,
      name: 'foo',
      key: 'bar',
      title: 'Title',
      description: 'description',
      fields: [{
        component: components.TEXT_FIELD,
        name: 'nested component',
        label: 'foo',
      }],
      additionalItems: {
        key: 1,
        name: 'foo',
        component: components.TEXT_FIELD,
        fields: [],
      },
    }];

    const wrapper = mount(
      <ContextWrapper schema={{ fields: formFields }}/>
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should extract dynamic field name if its in dot notation', () => {
    const formFields = [{
      component: components.FIELD_ARRAY,
      name: 'foo',
      key: 'bar',
      title: 'Title',
      description: 'description',
      fields: [{
        component: components.TEXT_FIELD,
        name: 'name.nested-component',
        label: 'foo',
      }],
    }];

    const wrapper = mount(
      <ContextWrapper initialValues={{ foo: [{ name: 'bar'  }]}} schema={{ fields: formFields }}/>
    );
    expect(wrapper.find({ name: 'foo[0]nested-component' })).toBeTruthy();
  });

  it('should extract dynamic field name if it contains items prefix', () => {
    const formFields = [{
      component: components.FIELD_ARRAY,
      name: 'foo',
      key: 'bar',
      title: 'Title',
      description: 'description',
      fields: [{
        component: components.TEXT_FIELD,
        name: 'items.nested-component',
        label: 'foo',
      }],
    }];

    const wrapper = mount(
      <ContextWrapper initialValues={{ foo: [{ name: 'bar'  }]  }} schema={{ fields: formFields }}/>
    );
    expect(wrapper.find({ name: 'foo[0]nested-component' })).toBeTruthy();
  });
});
