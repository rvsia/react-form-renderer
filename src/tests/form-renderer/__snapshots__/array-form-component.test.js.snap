// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderForm function should render array field correctly 1`] = `
<ContextWrapper>
  <ReactFinalForm
    initialValues={Object {}}
    mutators={
      Object {
        "insert": [Function],
        "move": [Function],
        "pop": [Function],
        "push": [Function],
        "remove": [Function],
        "shift": [Function],
        "swap": [Function],
        "unshift": [Function],
        "update": [Function],
      }
    }
    onSubmit={[MockFunction]}
  >
    <FormConditionWrapper>
      <FieldWrapper
        component={[Function]}
        componentType="field-array"
        fields={
          Array [
            Object {
              "component": "text-field",
              "label": "foo",
              "name": "nested component",
            },
          ]
        }
        formOptions={
          Object {
            "renderForm": [Function],
          }
        }
        key="bar"
        name="foo"
        validate={
          Array [
            undefined,
          ]
        }
      >
        <_class
          FieldProvider={[Function]}
          arrayValidator={[Function]}
          component={[Function]}
          fields={
            Array [
              Object {
                "component": "text-field",
                "label": "foo",
                "name": "nested component",
              },
            ]
          }
          formOptions={
            Object {
              "hasFixedItems": false,
              "renderForm": [Function],
            }
          }
          name="foo"
        >
          <Field
            FieldProvider={[Function]}
            arrayValidator={[Function]}
            component={[Function]}
            fields={
              Array [
                Object {
                  "component": "text-field",
                  "label": "foo",
                  "name": "nested component",
                },
              ]
            }
            formOptions={
              Object {
                "hasFixedItems": false,
                "renderForm": [Function],
              }
            }
            format={[Function]}
            name="foo"
            parse={[Function]}
            reactFinalForm={
              Object {
                "batch": [Function],
                "blur": [Function],
                "change": [Function],
                "focus": [Function],
                "getFieldState": [Function],
                "getRegisteredFields": [Function],
                "getState": [Function],
                "initialize": [Function],
                "isValidationPaused": [Function],
                "mutators": Object {
                  "insert": [Function],
                  "move": [Function],
                  "pop": [Function],
                  "push": [Function],
                  "remove": [Function],
                  "shift": [Function],
                  "swap": [Function],
                  "unshift": [Function],
                  "update": [Function],
                },
                "pauseValidation": [Function],
                "registerField": [Function],
                "reset": [Function],
                "resumeValidation": [Function],
                "setConfig": [Function],
                "submit": [Function],
                "subscribe": [Function],
              }
            }
          >
            <Component
              FieldProvider={[Function]}
              arrayValidator={[Function]}
              fields={
                Array [
                  Object {
                    "component": "text-field",
                    "label": "foo",
                    "name": "nested component",
                  },
                ]
              }
              formOptions={
                Object {
                  "hasFixedItems": false,
                  "renderForm": [Function],
                }
              }
              input={
                Object {
                  "name": "foo",
                  "onBlur": [Function],
                  "onChange": [Function],
                  "onFocus": [Function],
                  "value": "",
                }
              }
              meta={
                Object {
                  "active": false,
                  "data": Object {},
                  "dirty": false,
                  "dirtySinceLastSubmit": false,
                  "error": undefined,
                  "initial": undefined,
                  "invalid": false,
                  "pristine": true,
                  "submitError": undefined,
                  "submitFailed": false,
                  "submitSucceeded": false,
                  "submitting": false,
                  "touched": false,
                  "valid": true,
                  "visited": false,
                }
              }
            >
              <renderArrayField
                FieldProvider={[Function]}
                arrayValidator={[Function]}
                fields={
                  Array [
                    Object {
                      "component": "text-field",
                      "label": "foo",
                      "name": "nested component",
                    },
                  ]
                }
                formOptions={
                  Object {
                    "hasFixedItems": false,
                    "renderForm": [Function],
                  }
                }
                input={
                  Object {
                    "name": "foo",
                    "onBlur": [Function],
                    "onChange": [Function],
                    "onFocus": [Function],
                    "value": "",
                  }
                }
                meta={
                  Object {
                    "active": false,
                    "data": Object {},
                    "dirty": false,
                    "dirtySinceLastSubmit": false,
                    "error": undefined,
                    "initial": undefined,
                    "invalid": false,
                    "pristine": true,
                    "submitError": undefined,
                    "submitFailed": false,
                    "submitSucceeded": false,
                    "submitting": false,
                    "touched": false,
                    "valid": true,
                    "visited": false,
                  }
                }
                validate={Array []}
              >
                <DynamicArray
                  FieldProvider={[Function]}
                  arrayValidator={[Function]}
                  fieldKey="foo"
                  fields={
                    Array [
                      Object {
                        "component": "text-field",
                        "label": "foo",
                        "name": "nested component",
                      },
                    ]
                  }
                  formOptions={
                    Object {
                      "hasFixedItems": false,
                      "renderForm": [Function],
                    }
                  }
                  input={
                    Object {
                      "name": "foo",
                      "onBlur": [Function],
                      "onChange": [Function],
                      "onFocus": [Function],
                      "value": "",
                    }
                  }
                  meta={
                    Object {
                      "active": false,
                      "data": Object {},
                      "dirty": false,
                      "dirtySinceLastSubmit": false,
                      "error": undefined,
                      "initial": undefined,
                      "invalid": false,
                      "pristine": true,
                      "submitError": undefined,
                      "submitFailed": false,
                      "submitSucceeded": false,
                      "submitting": false,
                      "touched": false,
                      "valid": true,
                      "visited": false,
                    }
                  }
                  renderForm={[Function]}
                  validate={Array []}
                >
                  <_class
                    key="foo"
                    name="foo"
                    validate={[Function]}
                  >
                    <ReactFinalFormFieldArray(4.8.1)(2.0.1)
                      name="foo"
                      reactFinalForm={
                        Object {
                          "batch": [Function],
                          "blur": [Function],
                          "change": [Function],
                          "focus": [Function],
                          "getFieldState": [Function],
                          "getRegisteredFields": [Function],
                          "getState": [Function],
                          "initialize": [Function],
                          "isValidationPaused": [Function],
                          "mutators": Object {
                            "insert": [Function],
                            "move": [Function],
                            "pop": [Function],
                            "push": [Function],
                            "remove": [Function],
                            "shift": [Function],
                            "swap": [Function],
                            "unshift": [Function],
                            "update": [Function],
                          },
                          "pauseValidation": [Function],
                          "registerField": [Function],
                          "reset": [Function],
                          "resumeValidation": [Function],
                          "setConfig": [Function],
                          "submit": [Function],
                          "subscribe": [Function],
                        }
                      }
                      validate={[Function]}
                    >
                      <Col
                        xs={11}
                      >
                        <div />
                      </Col>
                      <Col
                        className="final-form-array-add-container"
                        xs={1}
                      >
                        <div>
                          <FormGroup>
                            <div>
                              <ButtonGroup
                                className="pull-right"
                              >
                                <div>
                                  <Button
                                    onClick={[Function]}
                                    type="button"
                                  >
                                    <button
                                      onClick={[Function]}
                                      type="button"
                                    />
                                  </Button>
                                </div>
                              </ButtonGroup>
                            </div>
                          </FormGroup>
                        </div>
                      </Col>
                    </ReactFinalFormFieldArray(4.8.1)(2.0.1)>
                  </_class>
                </DynamicArray>
              </renderArrayField>
            </Component>
          </Field>
        </_class>
      </FieldWrapper>
    </FormConditionWrapper>
  </ReactFinalForm>
</ContextWrapper>
`;

exports[`renderForm function should render fixed array field correctly 1`] = `
<ContextWrapper>
  <ReactFinalForm
    initialValues={Object {}}
    mutators={
      Object {
        "insert": [Function],
        "move": [Function],
        "pop": [Function],
        "push": [Function],
        "remove": [Function],
        "shift": [Function],
        "swap": [Function],
        "unshift": [Function],
        "update": [Function],
      }
    }
    onSubmit={[MockFunction]}
  >
    <FormConditionWrapper>
      <FieldWrapper
        additionalItems={
          Object {
            "component": "text-field",
            "fields": Array [],
            "key": 1,
            "name": "foo",
          }
        }
        component={[Function]}
        componentType="fixed-list"
        description="description"
        fields={
          Array [
            Object {
              "component": "text-field",
              "label": "foo",
              "name": "nested component",
            },
          ]
        }
        formOptions={
          Object {
            "renderForm": [Function],
          }
        }
        key="bar"
        name="foo"
        title="Title"
        validate={
          Array [
            undefined,
          ]
        }
      >
        <_class
          FieldProvider={[Function]}
          additionalItems={
            Object {
              "component": "text-field",
              "fields": Array [],
              "key": 1,
              "name": "foo",
            }
          }
          arrayValidator={[Function]}
          component={[Function]}
          description="description"
          fields={
            Array [
              Object {
                "component": "text-field",
                "label": "foo",
                "name": "nested component",
              },
            ]
          }
          formOptions={
            Object {
              "hasFixedItems": true,
              "renderForm": [Function],
            }
          }
          name="foo"
          title="Title"
        >
          <Field
            FieldProvider={[Function]}
            additionalItems={
              Object {
                "component": "text-field",
                "fields": Array [],
                "key": 1,
                "name": "foo",
              }
            }
            arrayValidator={[Function]}
            component={[Function]}
            description="description"
            fields={
              Array [
                Object {
                  "component": "text-field",
                  "label": "foo",
                  "name": "nested component",
                },
              ]
            }
            formOptions={
              Object {
                "hasFixedItems": true,
                "renderForm": [Function],
              }
            }
            format={[Function]}
            name="foo"
            parse={[Function]}
            reactFinalForm={
              Object {
                "batch": [Function],
                "blur": [Function],
                "change": [Function],
                "focus": [Function],
                "getFieldState": [Function],
                "getRegisteredFields": [Function],
                "getState": [Function],
                "initialize": [Function],
                "isValidationPaused": [Function],
                "mutators": Object {
                  "insert": [Function],
                  "move": [Function],
                  "pop": [Function],
                  "push": [Function],
                  "remove": [Function],
                  "shift": [Function],
                  "swap": [Function],
                  "unshift": [Function],
                  "update": [Function],
                },
                "pauseValidation": [Function],
                "registerField": [Function],
                "reset": [Function],
                "resumeValidation": [Function],
                "setConfig": [Function],
                "submit": [Function],
                "subscribe": [Function],
              }
            }
            title="Title"
          >
            <Component
              FieldProvider={[Function]}
              additionalItems={
                Object {
                  "component": "text-field",
                  "fields": Array [],
                  "key": 1,
                  "name": "foo",
                }
              }
              arrayValidator={[Function]}
              description="description"
              fields={
                Array [
                  Object {
                    "component": "text-field",
                    "label": "foo",
                    "name": "nested component",
                  },
                ]
              }
              formOptions={
                Object {
                  "hasFixedItems": true,
                  "renderForm": [Function],
                }
              }
              input={
                Object {
                  "name": "foo",
                  "onBlur": [Function],
                  "onChange": [Function],
                  "onFocus": [Function],
                  "value": "",
                }
              }
              meta={
                Object {
                  "active": false,
                  "data": Object {},
                  "dirty": false,
                  "dirtySinceLastSubmit": false,
                  "error": undefined,
                  "initial": undefined,
                  "invalid": false,
                  "pristine": true,
                  "submitError": undefined,
                  "submitFailed": false,
                  "submitSucceeded": false,
                  "submitting": false,
                  "touched": false,
                  "valid": true,
                  "visited": false,
                }
              }
              title="Title"
            >
              <renderArrayField
                FieldProvider={[Function]}
                additionalItems={
                  Object {
                    "component": "text-field",
                    "fields": Array [],
                    "key": 1,
                    "name": "foo",
                  }
                }
                arrayValidator={[Function]}
                description="description"
                fields={
                  Array [
                    Object {
                      "component": "text-field",
                      "label": "foo",
                      "name": "nested component",
                    },
                  ]
                }
                formOptions={
                  Object {
                    "hasFixedItems": true,
                    "renderForm": [Function],
                  }
                }
                input={
                  Object {
                    "name": "foo",
                    "onBlur": [Function],
                    "onChange": [Function],
                    "onFocus": [Function],
                    "value": "",
                  }
                }
                meta={
                  Object {
                    "active": false,
                    "data": Object {},
                    "dirty": false,
                    "dirtySinceLastSubmit": false,
                    "error": undefined,
                    "initial": undefined,
                    "invalid": false,
                    "pristine": true,
                    "submitError": undefined,
                    "submitFailed": false,
                    "submitSucceeded": false,
                    "submitting": false,
                    "touched": false,
                    "valid": true,
                    "visited": false,
                  }
                }
                title="Title"
                validate={Array []}
              >
                <FixedArrayField
                  FieldProvider={[Function]}
                  additionalItems={
                    Object {
                      "component": "text-field",
                      "fields": Array [],
                      "key": 1,
                      "name": "foo",
                    }
                  }
                  arrayValidator={[Function]}
                  description="description"
                  fields={
                    Array [
                      Object {
                        "component": "text-field",
                        "label": "foo",
                        "name": "nested component",
                      },
                    ]
                  }
                  formOptions={
                    Object {
                      "hasFixedItems": true,
                      "renderForm": [Function],
                    }
                  }
                  input={
                    Object {
                      "name": "foo",
                      "onBlur": [Function],
                      "onChange": [Function],
                      "onFocus": [Function],
                      "value": "",
                    }
                  }
                  meta={
                    Object {
                      "active": false,
                      "data": Object {},
                      "dirty": false,
                      "dirtySinceLastSubmit": false,
                      "error": undefined,
                      "initial": undefined,
                      "invalid": false,
                      "pristine": true,
                      "submitError": undefined,
                      "submitFailed": false,
                      "submitSucceeded": false,
                      "submitting": false,
                      "touched": false,
                      "valid": true,
                      "visited": false,
                    }
                  }
                  renderForm={[Function]}
                  title="Title"
                  validate={Array []}
                >
                  <Col
                    xs={12}
                  >
                    <div>
                      <h3>
                        Title
                      </h3>
                    </div>
                  </Col>
                  <Col
                    xs={12}
                  >
                    <div>
                      <p>
                        description
                      </p>
                    </div>
                  </Col>
                  <FormConditionWrapper>
                    <FieldWrapper
                      component={[Function]}
                      componentType="text-field"
                      formOptions={
                        Object {
                          "renderForm": [Function],
                        }
                      }
                      label="foo"
                      name="nested component"
                      validate={
                        Array [
                          undefined,
                        ]
                      }
                    >
                      <_class
                        FieldProvider={[Function]}
                        component={[Function]}
                        formOptions={
                          Object {
                            "renderForm": [Function],
                          }
                        }
                        label="foo"
                        name="nested component"
                        validate={[Function]}
                      >
                        <Field
                          FieldProvider={[Function]}
                          component={[Function]}
                          formOptions={
                            Object {
                              "renderForm": [Function],
                            }
                          }
                          format={[Function]}
                          label="foo"
                          name="nested component"
                          parse={[Function]}
                          reactFinalForm={
                            Object {
                              "batch": [Function],
                              "blur": [Function],
                              "change": [Function],
                              "focus": [Function],
                              "getFieldState": [Function],
                              "getRegisteredFields": [Function],
                              "getState": [Function],
                              "initialize": [Function],
                              "isValidationPaused": [Function],
                              "mutators": Object {
                                "insert": [Function],
                                "move": [Function],
                                "pop": [Function],
                                "push": [Function],
                                "remove": [Function],
                                "shift": [Function],
                                "swap": [Function],
                                "unshift": [Function],
                                "update": [Function],
                              },
                              "pauseValidation": [Function],
                              "registerField": [Function],
                              "reset": [Function],
                              "resumeValidation": [Function],
                              "setConfig": [Function],
                              "submit": [Function],
                              "subscribe": [Function],
                            }
                          }
                          validate={[Function]}
                        >
                          <Component
                            FieldProvider={[Function]}
                            formOptions={
                              Object {
                                "renderForm": [Function],
                              }
                            }
                            input={
                              Object {
                                "name": "nested component",
                                "onBlur": [Function],
                                "onChange": [Function],
                                "onFocus": [Function],
                                "value": "",
                              }
                            }
                            label="foo"
                            meta={
                              Object {
                                "active": false,
                                "data": Object {},
                                "dirty": false,
                                "dirtySinceLastSubmit": false,
                                "error": undefined,
                                "initial": undefined,
                                "invalid": false,
                                "pristine": true,
                                "submitError": undefined,
                                "submitFailed": false,
                                "submitSucceeded": false,
                                "submitting": false,
                                "touched": false,
                                "valid": true,
                                "visited": false,
                              }
                            }
                          >
                            <div
                              className="nested-item"
                            >
                              Text field
                            </div>
                          </Component>
                        </Field>
                      </_class>
                    </FieldWrapper>
                  </FormConditionWrapper>
                  <FormConditionWrapper>
                    <FieldWrapper
                      component={[Function]}
                      componentType="text-field"
                      fields={Array []}
                      formOptions={
                        Object {
                          "renderForm": [Function],
                        }
                      }
                      key="1"
                      name="foo"
                      validate={
                        Array [
                          undefined,
                        ]
                      }
                    >
                      <_class
                        FieldProvider={[Function]}
                        component={[Function]}
                        fields={Array []}
                        formOptions={
                          Object {
                            "renderForm": [Function],
                          }
                        }
                        name="foo"
                        validate={[Function]}
                      >
                        <Field
                          FieldProvider={[Function]}
                          component={[Function]}
                          fields={Array []}
                          formOptions={
                            Object {
                              "renderForm": [Function],
                            }
                          }
                          format={[Function]}
                          name="foo"
                          parse={[Function]}
                          reactFinalForm={
                            Object {
                              "batch": [Function],
                              "blur": [Function],
                              "change": [Function],
                              "focus": [Function],
                              "getFieldState": [Function],
                              "getRegisteredFields": [Function],
                              "getState": [Function],
                              "initialize": [Function],
                              "isValidationPaused": [Function],
                              "mutators": Object {
                                "insert": [Function],
                                "move": [Function],
                                "pop": [Function],
                                "push": [Function],
                                "remove": [Function],
                                "shift": [Function],
                                "swap": [Function],
                                "unshift": [Function],
                                "update": [Function],
                              },
                              "pauseValidation": [Function],
                              "registerField": [Function],
                              "reset": [Function],
                              "resumeValidation": [Function],
                              "setConfig": [Function],
                              "submit": [Function],
                              "subscribe": [Function],
                            }
                          }
                          validate={[Function]}
                        >
                          <Component
                            FieldProvider={[Function]}
                            fields={Array []}
                            formOptions={
                              Object {
                                "renderForm": [Function],
                              }
                            }
                            input={
                              Object {
                                "name": "foo",
                                "onBlur": [Function],
                                "onChange": [Function],
                                "onFocus": [Function],
                                "value": "",
                              }
                            }
                            meta={
                              Object {
                                "active": false,
                                "data": Object {},
                                "dirty": false,
                                "dirtySinceLastSubmit": false,
                                "error": undefined,
                                "initial": undefined,
                                "invalid": false,
                                "pristine": true,
                                "submitError": undefined,
                                "submitFailed": false,
                                "submitSucceeded": false,
                                "submitting": false,
                                "touched": false,
                                "valid": true,
                                "visited": false,
                              }
                            }
                          >
                            <div
                              className="nested-item"
                            >
                              Text field
                            </div>
                          </Component>
                        </Field>
                      </_class>
                    </FieldWrapper>
                  </FormConditionWrapper>
                </FixedArrayField>
              </renderArrayField>
            </Component>
          </Field>
        </_class>
      </FieldWrapper>
    </FormConditionWrapper>
  </ReactFinalForm>
</ContextWrapper>
`;
